File: src/modules/algorithm/seed/algorithms/maze-solver/languages/javascript/dfs-solution.js
/**
 * Solves a maze using Depth-First Search (DFS).
 *
 * @param {number[][]} maze - The maze represented as a 2D array.
 * @param {number[]} start - The starting point in the maze [row, col].
 * @param {number[]} end - The ending point in the maze [row, col].
 * @returns {number[][]} The path from start to end as an array of [row, col] coordinates, or null if no solution exists.
 */
function solveMaze(maze, start, end) {
  const rows = maze.length;
  const cols = maze[0].length;
  const path = [];
  const visited = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(false));

  function dfs(row, col) {
    if (
      row < 0 ||
      row >= rows ||
      col < 0 ||
      col >= cols ||
      maze[row][col] === 1 ||
      visited[row][col]
    ) {
      return false;
    }

    visited[row][col] = true;
    path.push([row, col]);

    if (row === end[0] && col === end[1]) {
      return true;
    }

    // Explore in all four directions
    if (
      dfs(row - 1, col) || // Up
      dfs(row + 1, col) || // Down
      dfs(row, col - 1) || // Left
      dfs(row, col + 1)
    ) {
      // Right
      return true;
    }

    // Backtrack if no solution is found
    path.pop();
    return false;
  }

  if (dfs(start[0], start[1])) {
    return path;
  }

  return null;
}

module.exports = { solveMaze };

File: src/modules/algorithm/seed/algorithms/maze-solver/languages/python/dfs-solution.py
from typing import List, Tuple

Maze = List[List[int]]
Point = Tuple[int, int]

def solve_maze(maze: Maze, start: Point, end: Point) -> List[Point] | None:
    """
    Solves a maze using Depth-First Search (DFS).

    Args:
        maze: The maze represented as a 2D list.
        start: The starting point in the maze (row, col).
        end: The ending point in the maze (row, col).

    Returns:
        The path from start to end as a list of (row, col) coordinates, or None if no solution exists.
    """
    rows = len(maze)
    cols = len(maze[0])
    path: List[Point] = []
    visited: List[List[bool]] = [[False for _ in range(cols)] for _ in range(rows)]

    def dfs(row: int, col: int) -> bool:
        if row < 0 or row >= rows or col < 0 or col >= cols or maze[row][col] == 1 or visited[row][col]:
            return False

        visited[row][col] = True
        path.append((row, col))

        if (row, col) == end:
            return True

        # Explore in all four directions
        if (dfs(row - 1, col) or  # Up
            dfs(row + 1, col) or  # Down
            dfs(row, col - 1) or  # Left
            dfs(row, col + 1)):   # Right
            return True

        # Backtrack if no solution is found
        path.pop()
        return False

    if dfs(start[0], start[1]):
        return path
    return None 